tools/wrapper/retrace.c

```diff
@@ -0,0 +1,93 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <errno.h>
+#include <unistd.h>
+#include <sys/types.h>
+#include <sys/wait.h>
+
+/*
+ * print usage
+ */
+
+static void usage(const char *prog_name)
+{
+	fprintf(stderr, "Usage: %s -l <path of libretrace.so> <command line>\n",
+			prog_name);
+	exit(1);
+}
+
+/*
+ * fork process
+ */
+
+static void fork_process(int argc, char **argv)
+{
+	pid_t pid;
+	int dup_stdout, dup_stderr;
+
+	char **args;
+	int i;
+
+	/* set arguments */
+	args = (char **) malloc((argc - 1) * sizeof(char **));
+	if (!args) {
+		fprintf(stderr, "Out of memory!\n");
+		exit(1);
+	}
+
+	for (i = 0; i < argc - 3; i++)
+		args[i] = argv[i + 3];
+
+	/* open stdout */
+	dup2(STDOUT_FILENO, dup_stdout);
+	dup2(STDOUT_FILENO, dup_stderr);
+
+	/* fork new process */
+	pid = fork();
+	if (pid == 0) {
+		dup2(STDOUT_FILENO, dup_stdout);
+		dup2(STDERR_FILENO, dup_stderr);
+
+		/* set environment for preload */
+		setenv("LD_PRELOAD", argv[2], 1);
+		exit(execvp(argv[3], args));
+	} else if (pid < 0) {
+		fprintf(stderr, "Failed to fork new process(err:%d)\n", errno);
+		exit(1);
+	}
+
+	/* wait until the process has terminated */
+	do {
+		int w, status;
+
+		/* wait until process has terminated */
+		w = waitpid(pid, &status, -1);
+		if (w < 0) {
+			fprintf(stderr, "waitpid() error for command\n");
+			break;
+		}
+
+		/* set exit status */
+		if (WIFEXITED(status))
+			break;
+	} while (1);
+}
+
+/*
+ * main function
+ */
+
+int main(int argc, char *argv[])
+{
+	/* check arguments */
+	if (argc < 4)
+		usage(argv[0]);
+
+	if (strcmp(argv[1], "-l") != 0)
+		usage(argv[0]);
```

Comment:
Rather use `getopt` here and use default locations for libretrace (probably /use/lib, /usr/local/lib) in case `-l <soname>` is not specified.

Reply:
@drystone Using `getopt` function is some complex because it needs to specify `command line`  to be executed.

Reply:
This is a very common use case for `getopt` which is only interested in the option characters (and/or option names for `getopt_long`) and their respective arguments. The [man page](https://linux.die.net/man/3/getopt) has all the info.

Reply:
Yes, I know about `getopt` function. But this utility has command line options to be executed.
It's not possible to be parsed by `getopt` properly.
If the syntax is specified in the following format, then using of 'getopt' is possible:

```
./retrace -l <library> -c <"command line">
```
In this case, the command line should be in `" "`. I think it's not preferred way.

Reply:
```
#include <unistd.h>
#include <stdio.h>

void
main(int argc, char *argv[])
{
        for (;;) {
                switch (getopt(argc, argv, "a::b:c")) {
                case 'a':
                        printf("getopt a: %s\n", optarg ? optarg : NULL);
                        break;
                case 'b':
                        printf("getopt b: %s\n", optarg);
                        break;
                case 'c':
                        printf("getopt c: (flag)\n");                    
                        break;
                default:
                        goto done_options;
                }
        }

done_options:
        while (optind < argc) {
                printf("argument: %s\n", argv[optind]);
                ++optind;
        }
}
```
```
$ make t
cc    -c -o t.o t.c
cc   t.o   -o t
$./t -aoptional -bmandatory -c execute me please
getopt a: optional
getopt b: mandatory
getopt c: (flag)
argument: execute
argument: me
argument: please
```

Reply:
@drystone Assume the following example:
The first example:
```
./wrapper -l /bin/ls
```

The second example:
```
./wrapper /bin/ls -l
```
For both cases, `getopt` process `-l` option as `wrapper` utility.
But in the second case, `-l` option is for `/bin/ls`.

That's why I don't use `getopt` when parsing command line arguments.

Reply:
There is the '+' mode for this. 
```
By default, getopt() permutes  the  con‐
tents of argv as it scans, so that even‐
tually all the  nonoptions  are  at  the
end.   Two  other  modes are also imple‐
mented.  If the first character of  opt‐
string  is  '+' or the environment vari‐
able POSIXLY_CORRECT is set, then option
processing  stops as soon as a nonoption
argument is encountered.
```